src/server/dflycmd.cc

```diff
@@ -380,79 +372,105 @@ void DflyCmd::FullSyncFb(FlowInfo* flow) {
   }
 
   if (ec) {
-    LOG(ERROR) << ec;
-    return;
+    return cntx->Error(ec);
   }
 
-  // TODO: we should be able to stop earlier if requested.
-  ec = saver->SaveBody(nullptr);
-  if (ec) {
-    LOG(ERROR) << ec;
-    return;
+  if ((ec = saver->SaveBody(*cntx, nullptr))) {
+    return cntx->Error(ec);
   }
 
-  VLOG(1) << "Sending full sync EOF";
-
   ec = flow->conn->socket()->Write(io::Buffer(flow->eof_token));
   if (ec) {
-    LOG(ERROR) << ec;
-    return;
+    return cntx->Error(ec);
   }
 }
 
 uint32_t DflyCmd::CreateSyncSession() {
   unique_lock lk(mu_);
+  unsigned sync_id = next_sync_id_++;
 
-  auto sync_info = make_shared<SyncInfo>();
-  sync_info->flows.resize(shard_set->size() + 1);
+  unsigned flow_count = shard_set->size() + 1;
+  auto err_handler = [this, sync_id](const GenericError& err) {
+    LOG(INFO) << "Running error handler for sync session " << sync_id;
 
-  auto [it, inserted] = sync_infos_.emplace(next_sync_id_, std::move(sync_info));
+    // Stop replication in case of error.
+    // StopReplication needs to run async to prevent blocking
+    // the error handler.
+    ::boost::fibers::fiber{&DflyCmd::StopReplication, this, sync_id}.detach();
```

Comment:
I successfully stripped out the third lock for the cancellation/connection on-hold phase.

It seems to work, _but_ _its really easy_ to deadlock or stall in some places.

Besides, the absence of fast-fail might still be a problem, but we'll see with the replica code. The question is basically how to distribute unblocking complexity between the two actors

Reply:
do we have a test/scenario that deadlocks?

If we are reasonably stable (i.e. with a single replica we work reliably in 99.99% of the time - we can fix the issues later).  For now - a single replica is the major use-case - to support server migrations.